doc/man-source/irsend.1
.TH IRSEND "1" "Last change: Mar 2017" "irsend @version@" "User Commands"
.SH NAME
.P
\fBirsend\fR - basic LIRC program to send infra-red commands
.SH SYNOPSIS
.P
\fBirsend\fR [\fIoptions\fR] \fIsend_once \fI<remote>\fR \fI<code>\fR [\fIcode...]\fR
.br
\fBirsend\fR [\fIoptions\fR] \fIsend_start \fI<remote>\fR \fI<code>\fR
.br
\fBirsend\fR [\fIoptions\fR] \fIsend_stop \fI<remote>\fR \fI<code>\fR
.br
\fBirsend\fR [\fIoptions\fR] \fIlist\fR \fI<remote>\fR \fI<code>\fR
.br
\fBirsend\fR [\fIoptions\fR] \fIset_transmitters\fR \fI<num>\fR \fI[num...]\fR
.br
\fBirsend\fR [\fIoptions\fR] \fIsimulate\fR \fI<button press packet>\fR
.SH DESCRIPTION
.P
Asks the \fBlircd\fR daemon to send one or more CIR
(Consumer Infra-Red) commands.
This is intended for remote control of electronic devices such as
TV boxes, HiFi sets, etc.
.PP
\fBirsend\fR supports six sub-commands:
.nf
\fBsend_once\fR         - send one or more code(s) once
\fBsend_start\fR        - start repeating a code.
\fBsend_stop\fR         - stop repeating code.
\fBlist\fR              - list configured remote items
\fBset_transmitters\fR  - set active transmitters
\fBsimulate\fR          - simulate IR event
.fi
.P
\fIremote\fR is the name of a remote, as described in the \fBlircd\fR
configuration file.
.P
\fIcode\fR is the name of a remote control key of \fIremote\fR, as it
appears in the \fBlircd.conf\fR configuration file.
.P
\fInum\fR is the transmitter number of the hardware device.
.P
For the \fBlist\fR command, \fIremote\fR and/or \fIcode\fR
can be empty:
.IP
.nf
\fBlist\fR   ""    ""   - list all configured remote names
\fBlist\fR \fIremote\fR  ""   - list all codes of \fIremote\fR
\fBlist\fR \fIremote\fR \fIcode\fR  - list only \fIcode\fR of \fIremote\fR
.fi
.P
The \fBsimulate\fR command only works if it has been explicitly
enabled in lircd using the --allow-simulate option.
The required \fIbutton press packet\fR should formatted as a socket
broadcast message documented in \fBlircd(8)\fR.
.SH OPTIONS
.TP
\fB\-h\fR \fB\-\-help\fR
Display usage summary.
.TP
\fB\-v\fR \fB\-\-version\fR
Display version.
.TP
\fB\-d\fR \fB\-\-device\fR=\fIdevice\fR
Use given lircd socket [@varrundir@/lirc/lircd].
.TP
\fB\-a\fR \fB\-\-address\fR=\fIhost[\fR:port]
Connect to lircd at this address.
.TP
\-# \fB\-\-count\fR=\fIn\fR
Send command n times.

.SH ENVIRONMENT
.TP 4
LIRC_SOCKET_PATH
The lircd socket to connect to, defaults to a hardcoded default value
@varrundir@/lirc/lircd
.SH FILES
.TP
.I @etcdir@/lirc/lircd.conf
Default \fBlircd\fR configuration file.  It should contain all the
remotes, their infra-red codes and the corresponding timing and
waveform details.

.SH DIAGNOSTICS
.P
If \fBlircd\fR is not running (or @varrundir@/lirc/lircd lacks write
permissions) \fBirsend\fR aborts with the following diagnostics:
.nf
"irsend: could not connect to socket"
"irsend: Connection refused" (or "Permission denied").
.fi
.SH EXAMPLES
.nf
irsend LIST DenonTuner ""
irsend SEND_ONCE  DenonTuner PROG\-SCAN
irsend SEND_ONCE  OnkyoAmpli VOL\-UP VOL\-UP VOL\-UP VOL\-UP
irsend SEND_START OnkyoAmpli VOL\-DOWN ; sleep 3
irsend SEND_STOP  OnkyoAmpli VOL\-DOWN
irsend SET_TRANSMITTERS 1
irsend SET_TRANSMITTERS 1 3 4
irsend SIMULATE "0000000000000476 00 OK TECHNISAT_ST3004S"
.fi
.SH "DRIVER LOADING"
Drivers are loaded dynamically. The directory used for this is determined by (falling
priority):
.IP \- 2
The environment variable LIRC_PLUGINDIR.
.IP \- 2
The 'plugindir' entry in  the [lircd] section of the lirc_options.conf file.
.IP \- 2
A hardcoded default (@libpath@/lirc/plugins).
.SH "SEE ALSO"
lircd(8),
.br
mode2(1),
.br
xmode2(1),
.br
irrecord(1),
.br
irw(1)


NEWS
0.10.0-devel
 * The deprecated lircd --uinput option is dropped. Use
   the lircd-uinput(8) daemon instead.
 * The slinke (a. k. a. slink-e) driver is dropped.
 * plugins: Add support for irdroid in the irtoy driver.
 * Enhanced device enumeration, mode2 --list-devices works for
   most drivers.
 * New, shiny (i. e., unstable) python API. See the API docs,
   module "python bindings". The API is also available as a pypi
   package, typically at /usr/share/lirc/python-pkg
 * The configs/*conf files no more supports  enumerating devices
   using shell scripts.
 * Added enumeration support and server version level check to
   driver API.
 * Added manual devinput configuration, cross-compile support (#264).
 * lircd, lircd-uinput: Change default release suffix due to linux
   kernel changes (#263).

0.9.4d 23/01/17
 * The debian packaging means that freebsd and hurd builds are
   regularly tested, improving build stability.
 * A multitude of build patches, mostly after building on Debian:
    + Cross-compilation FTBS fixed (#233)
    + Fix configuration of configure (#234)
    + Fix library specification in lirc.pc (#236)
    + use HAVE_UINPUT=1 to force building uinput code (#238).
    + lib: Only use local headers in lirc_private.h (#237).
    + Configure libudev (#242)
    + Various kfreebsd fixes
    + Include config.h in installed headers (#250).
    + Fix _XOPEN_SOURCE related FTBS on some platforms (#248)
 * lib: Fix some lircrcd-related bugs (#254)
 * systemd: Fix wrong path to irexec (#239)
 * systemd: Add missing lircd-setup.service to distribution (#243)
 * lircd: Fix --listen option parsing [#249]
 * lircd: Fix segfault in SET_INPUTLOG without arguments (#252)
 * lircd: Fix DRV_OPTION error message handling (#256).
 * ircat: Fix option parsing bug + empty conditions (#251)
 * xmode2: Fix handling of --driver option.

0.9.4c 22/10/16
 * New curl_poll files adds a new MIT license. Lirc now carries
   GPL, BSD and MIT licenses.
 * Several patches to make lirc work again on macos (#226, #227).
 * New tool irtext2udp generates udp plugin data (#197).
 * Fix FTBS errors on linux kernel >= 4.8.0 (#232).
 * plugins: girs: Add udp/receive support (#201).
 * mode2: Fix broken --raw option (#223).
 * lirc-make-devinput: Fix that two remotes has the same name (#222).
 * lircd: Warn for duplicate remote definitions (#222).
 * lircd: Retry temporary unavailable write sockets (#221).
 * contrib: Fix usb devices' acl permissions (rhbz #1364744).
 * lirc-setup: Fix crasher bugfix (Fedora abort trace).
 * contrib: disable udev /dev/uinput rule requiring lirc user.
 * plugins: audio_alsa: Fix byte truncating in 16-bit data (#218):
 * plugins: irtoy: improve error handling (#220).
 * Minor documentation fixes.

0.9.4b 9/8/16
  * lircd-uinput: A multitude of fixes (#213, #211, #177). Adds new options.
  * lirc_client:  Fix "Can't redirect irsend output" (#207).
  * lirc_client: read_string() not terminating buffer contents (#216)
  * Remove a single python2 dependency [4e3c97c].
  * irrecord: Fix under-sized buffer causing crash (#208).
  * systemd: Add hardening options comments to services (#204).
  * lircmd: Options, udev examples and manpage update.A
  * lib: Add missing Makefile.am dep (#210).
  * doc: Manpages and configuration guide fixes.
  * build: Don't define docdir - autoconf defines docdir (#215).
  * build: support runstatedir configure option (#212).

0.9.4a 22/6/16
  * Fix syntax errors in systemd unit files (rhbz: 1350750).
  * irrecord: Enhance "Don't use devinput" message (#180).
  * plugins/uirt2_raw: Reset device to UIR mode on deinit.
  * plugins/girs.c: Enable reception after sending.
  * lirc.org: Updates and bugfixes
  * lircd, liblirc_client: Fix freeaddrinfo handling (#195).
  * Don't build commandir unless we have usb.h (#191).
  * lirc_options.conf: debug => loglevel (#199).
  * lib: use proper linking method to avoid parallel build issue
  * tools: make_rel_symlink.py can also use python2
  * plugins/devinput: Make the list-devices support available.
  * configure.ac: Bugfix (BSD setting reported wrong).

0.9.4 22/5/16
  * Issues: https://sf.net/p/lirc/tickets/milestone/0.9.4
  * Digest of 0.9.4pre1, 0.9.4pre2

  * The caraca driver has been retired (still available in attic/).
  * The iguanair driver has been moved out of tree and is now
    maintained by Iguanaworks, Inc. See drivers/iguanair/README.
  * The irman driver has been moved out of tree. From 0.5.1 the
    libirman package contains the driver. See drivers/irman/README
  * The yard2 driver has been moved out of tree and is now maintained
    by the yard2 project. See driver/yard2/README.
  * The lirc.4 manpage (/dev/lirc* interface) is upstreamed to the
    linux kernel manpages project (#164). Old page lirc.4l still
    lingers in attic/ but is not installed.
  * Upstream selinux updated to match current lircd requirements.
    See: https://github.com/fedora-selinux/selinux-policy/pull/41

  * New upstream debian packaging with which Debian users can create
    .deb packages. Ubuntu packages are also available in PPAs. See
    README.md in the debian branch (#137).
  * New driver girs for the GIRS protocol, gives LIRC Arduino support.
    See new info in driver documentation.
  * New transmit-only FTDI driver with enhanced accuracy and hardware
    support. See ftdix driver documentation.
  * Support for non-modulated signals added (#132).
  * As shipped, lircd now uses the devinput driver and the new 'auto'
    device, should work for all remotes supported by kernel (#106).
  * Enhanced lirc-setup tool (#123, partly):
    - Add drvctl device enumeration support.
    - Integrated mode2 and irw tests of current configuration.
    - Documentation links.
    - More consistent data based on new database.
    - Multiple bugfixes.
  * Several patches to make mode2 work with all drivers(#123, partly)
  * Configuration database overhaul (#123, partly)
    - Database lives as yaml files in configs/.

tools/irsend.cpp
/*

  irsend -  application for sending IR-codes via lirc

  Copyright (C) 1998 Christoph Bartelmus (lirc@bartelmus.de)

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA

*/

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include <getopt.h>
#include <errno.h>
#include <limits.h>

#include "lirc_log.h"
#include "lirc_client.h"

static const logchannel_t logchannel = LOG_APP;

static const char* const help =
	"\nSynopsis:\n"
	"    irsend [options] SEND_ONCE remote code [code...]\n"
	"    irsend [options] SEND_START remote code\n"
	"    irsend [options] SEND_STOP remote code\n"
	"    irsend [options] LIST remote\n"
	"    irsend [options] SET_TRANSMITTERS remote num [num...]\n"
	"    irsend [options] SIMULATE \"scancode repeat keysym remote\"\n"
	"Options:\n"
	"    -h --help\t\t\tdisplay usage summary\n"
	"    -v --version\t\tdisplay version\n"
	"    -d --device=device\t\tuse given lircd socket [" LIRCD "]\n"
	"    -a --address=host[:port]\tconnect to lircd at this address\n"
	"    -# --count=n\t\tsend command n times\n";

const char* prog;

int send_packet(lirc_cmd_ctx* ctx, int fd)
{
	int r;

	do {
		r = lirc_command_run(ctx, fd);
		if (r != 0 && r != EAGAIN)
			fprintf(stderr,
				"Error running command: %s\n", strerror(r));
	} while (r == EAGAIN);
	return r == 0 ? 0 : -1;
}

void reformat_simarg(char* code, char buffer[])
{
	unsigned int scancode;
	unsigned int repeat;
	char keysym[32];
	char remote[64];
	char trash[32];
	int r;

	r = sscanf(code, "%x %x %32s %64s %32s",
		   &scancode, &repeat, keysym, remote, trash);
	if (r != 4) {
		fprintf(stderr, "Bad simulate argument: %s\n", code);
		exit(EXIT_FAILURE);
	}
	snprintf(buffer, PACKET_SIZE, "%016x %02x %s %s",
		 scancode, repeat, keysym, remote);
}





int main(int argc, char** argv)
{
	char* directive;
	char* remote;
	char* code;
	const char* lircd = NULL;
	char* address = NULL;
	unsigned short port = LIRC_INET_PORT;
	unsigned long count = 1;
	int fd;
	char buffer[PACKET_SIZE + 1];
	int r;
	lirc_cmd_ctx ctx;

	prog = "irsend";

	while (1) {
		int c;
		static struct option long_options[] = {
			{ "help",    no_argument,	NULL, 'h' },
			{ "version", no_argument,	NULL, 'v' },
			{ "device",  required_argument, NULL, 'd' },
			{ "address", required_argument, NULL, 'a' },
			{ "count",   required_argument, NULL, '#' },
			{ 0,	     0,			0,    0	  }
		};
		c = getopt_long(argc, argv, "hvd:a:#:", long_options, NULL);
		if (c == -1)
			break;
		switch (c) {
		case 'h':
			fputs(help, stdout);
			return EXIT_SUCCESS;
		case 'v':
			printf("%s %s\n", prog, VERSION);
			return EXIT_SUCCESS;
		case 'd':
			lircd = optarg;
			break;
		case 'a':
		{
			char* p;
			char* end;
			unsigned long val;

			address = strdup(optarg);
			if (!address) {
				fprintf(stderr, "%s: out of memory\n", prog);
				return EXIT_FAILURE;
			}
			p = strchr(address, ':');
			if (p != NULL) {
				val = strtoul(p + 1, &end, 10);
				if (!(*(p + 1)) || *end || val < 1 || val > USHRT_MAX) {
					fprintf(stderr, "%s: invalid port number: %s\n", prog, p + 1);
					return EXIT_FAILURE;
				}
				port = (unsigned short)val;
				*p = 0;
			}
			break;
		}
		case '#':
		{
			char* end;

			count = strtoul(optarg, &end, 10);
			if (!*optarg || *end) {
				fprintf(stderr, "%s: invalid count value: %s\n", prog, optarg);
				return EXIT_FAILURE;
			}
			break;
		}
		default:
			return EXIT_FAILURE;
		}
	}
	if (optind + 2 > argc) {
		fprintf(stderr, "%s: not enough arguments\n", prog);
		return EXIT_FAILURE;
	}
	if (lircd == NULL)
		lircd = LIRCD;
	if (address == NULL)
		fd = lirc_get_local_socket(lircd ? lircd : NULL, 0);
	else
		fd = lirc_get_remote_socket(address, port, 0);
	if (fd < 0) {
		lircd = lircd ? lircd : getenv("LIRC_SOCKET_PATH");
		lircd = lircd ? lircd : LIRCD;
		perrorf("Cannot open socket %s", lircd);
		exit(EXIT_FAILURE);
	}
	if (address)
		free(address);
	address = NULL;

	directive = argv[optind++];

	if (strcasecmp(directive, "set_transmitters") == 0) {
		code = argv[optind++];
		if (strlen(directive) + strlen(code) + 2 < PACKET_SIZE) {
			sprintf(buffer, "%s %s", directive, code);
		} else {
			fprintf(stderr, "%s: input too long\n", prog);
			exit(EXIT_FAILURE);
		}
		while (optind < argc) {
			code = argv[optind++];
			if (strlen(buffer) + strlen(code) + 2 < PACKET_SIZE) {
				sprintf(buffer + strlen(buffer), " %s", code);
			} else {
				fprintf(stderr, "%s: input too long\n", prog);
				exit(EXIT_FAILURE);
			}
		}
		strcat(buffer, "\n");
		lirc_command_init(&ctx, buffer);
		if (send_packet(&ctx, fd) == -1)
			exit(EXIT_FAILURE);
	}
	if (strcasecmp(directive, "simulate") == 0) {
		code = argv[optind++];
		if (optind != argc) {
			fprintf(stderr, "%s: invalid argument count\n", prog);
			exit(EXIT_FAILURE);
		}
		reformat_simarg(code, buffer);
		r = lirc_command_init(&ctx, "%s %s\n", directive, buffer);
		if (r != 0) {
			fprintf(stderr, "%s: %s\n", prog, strerror(r));
			exit(EXIT_FAILURE);
		}
		if (send_packet(&ctx, fd) == -1)
			exit(EXIT_FAILURE);
	} else {
		remote = argv[optind++];

		if (optind == argc) {
			fprintf(stderr, "%s: not enough arguments\n", prog);
			exit(EXIT_FAILURE);
		}
		while (optind < argc) {
			code = argv[optind++];
			if (strcasecmp(directive, "SEND_ONCE") == 0 && count > 1)
				r = lirc_command_init(&ctx, "%s %s %s %lu\n",
						      directive, remote, code, count);
			else
				r = lirc_command_init(&ctx, "%s %s %s\n",
						      directive, remote, code);
			if (r != 0) {
				fprintf(stderr, "%s: input too long\n", prog);
				exit(EXIT_FAILURE);
			}
			lirc_command_reply_to_stdout(&ctx);
			if (send_packet(&ctx, fd) == -1)
				exit(EXIT_FAILURE);
		}
	}
	close(fd);
	return EXIT_SUCCESS;
}
